service-map/src/main/java/fi/nls/oskari/map/layer/formatters/LayerJSONFormatterUSERLAYER.java

```diff
@@ -34,49 +27,50 @@ public JSONObject getJSON(final OskariLayer baseLayer, UserLayer ulayer, String
-            arr.put(id);
-            for(int i = 0; i < fields.length(); i++){
-                JSONObject obj = fields.getJSONObject(i);
-                String name = obj.getString("name");
-                // skip geometry
-                if (DEFAULT_GEOMETRY_NAME.equals(name)) {
-                    continue;
-                }
-                JSONObject locales = obj.optJSONObject("locales");
-                String locale = name;
-                if (locales != null) {
-                    if (locales.has(lang)) {
-                        locale = locales.getString(lang);
-                    } else if (locales.has(DEFAULT_LOCALES_LANGUAGE)){
-                        locale = locales.getString(DEFAULT_LOCALES_LANGUAGE);
-                    }
-                }
-                JSONObject prop = new JSONObject();
-                prop.put(KEY_NAME, name);
-                prop.put(KEY_LOCALE, locale);
-                arr.put(prop);
+    // parse fields like WFSLayerAttributes
+    private static void parseFields(JSONObject layerJson, final JSONArray fields) {
+        if (fields == null || fields.length() == 0){
+            return;
+        }
+        JSONObject attributes = JSONHelper.getJSONObject(layerJson, "attributes");
+        JSONObject data = new JSONObject();
+        JSONObject types = new JSONObject();
+        JSONObject filter = new JSONObject();
+        Map<String, JSONObject> locale = new HashMap<>();
+        JSONArray filteredFields = new JSONArray();
+        for(int i = 0; i < fields.length(); i++){
+            JSONObject field = JSONHelper.getJSONObject(fields, i);
+            String name = JSONHelper.optString(field, "name");
+            String type = JSONHelper.optString(field, "type");
+            if (DEFAULT_GEOMETRY_NAME.equals(name)) {
+                JSONHelper.putValue(data, "geometryField", name);
+                JSONHelper.putValue(data, "geometryType", type);
+                continue;
             }
-            return arr;
-        } catch (Exception e) {
-            throw new IllegalArgumentException("Couldn't create locales JSONArray from fields");
+            // add name to filtered fields to order/filter fields
+            filteredFields.put(name);
+            // locales
+            Map<String,String> locales = JSONHelper.getObjectAsMap(field.optJSONObject("locales"));
+            locales.keySet().forEach(lang -> {
+                locale.putIfAbsent(lang, new JSONObject());
+                JSONHelper.putValue(locale.get(lang), name, locales.get(lang));
+            });
+            // attribute types, like WFSGetLayerFields
+            JSONHelper.putValue(types, name, WFSConversionHelper.getSimpleType(type.toLowerCase()));
         }
+
+        JSONHelper.putValue(attributes, "data", data);
+        JSONHelper.putValue(data, "filter", filter);
+        JSONHelper.putValue(data, "types", types);
+        JSONHelper.put(filter, "default", filteredFields);
```

Comment:
This could be `lang` instead of `default` since it's always the same we can just use the current users language (or the instances default lang is fine too).

Reply:
Removed `default` and changed `filter` to be array